conda/cli/main_clean.py

```diff
@@ -16,25 +17,57 @@
 _EXTS = (*CONDA_PACKAGE_EXTENSIONS, *(f"{e}.part" for e in CONDA_PACKAGE_EXTENSIONS))
 
 
-def find_tarballs():
+def _getsize(*parts: str, warnings: Optional[List[Tuple[str, Exception]]] = None) -> int:
+    path = join(*parts)
+    try:
+        stat = lstat(path)
+    except OSError as e:
+        if warnings is None:
+            raise
+        warnings.append((path, e))
+
+    # TODO: This doesn't handle packages that have hard links to files within
+    # themselves, like bin/python3.3 and bin/python3.3m in the Python package
+    if stat.st_nlink > 1:
+        raise NotImplementedError
+
+    return stat.st_size
+
+
+def find_tarballs() -> Dict[str, Any]:
     from ..core.package_cache_data import PackageCacheData
 
-    pkgs_dirs = {}
-    total_size = 0
+    warnings: List[Tuple[str, Exception]] = []
+    pkg_sizes: Dict[str, Dict[str, int]] = {}
     for package_cache in PackageCacheData.writable_caches(context.pkgs_dirs):
         pkgs_dir = package_cache.pkgs_dir
         if not isdir(pkgs_dir):
             continue
-        root, _, filenames = next(walk(pkgs_dir))
-        for fn in filenames:
-            if fn.endswith(_EXTS):
-                pkgs_dirs.setdefault(pkgs_dir, []).append(fn)
-                total_size += getsize(join(root, fn))
 
-    return {"pkgs_dirs": pkgs_dirs, "total_size": total_size}
+        # tarballs are files in pkgs_dir
+        _, _, tars = next(walk(pkgs_dir))
+        for tar in tars:
+            # tarballs also end in .tar.bz2, .conda, .tar.bz2.part, or .conda.part
+            if not tar.endswith(_EXTS):
+                continue
 
+            # get size
+            try:
+                size = _getsize(pkgs_dir, tar, warnings=warnings)
+            except NotImplementedError:
+                pass
+            else:
+                pkg_sizes.setdefault(pkgs_dir, {})[tar] = size
 
-def rm_tarballs(args, pkgs_dirs, total_size, verbose=True):
+    return {
+        "warnings": warnings,
+        "pkg_sizes": pkg_sizes,
+        "pkgs_dirs": {pkgs_dir: tuple(pkgs) for pkgs_dir, pkgs in pkg_sizes.items()},
+        "total_size": sum(sum(pkgs.values()) for pkgs in pkg_sizes.values()),
```

Comment:
I wonder if this would be good to have in an own private function like `_sum_sizes` since there is the same thing happening in `find_pkgs`?